### PR TITLE
New discard per subject feature blog post

### PR DESCRIPTION
**Pull request by jnmoyne**
Blog post about the new 'discard new per subject' funcitonality introduced in 2.9, a practical example of what you can do with it with infinite message deduplication, how it works, how it goes beyond what JS already does and what Kafka or any other distribute log streaming system can do.

### CODE REVIEW

**Comment by derekcollison** [comment]
Looping in @ripienaar as well since he uses this alot.

**Comment by ripienaar** [comment]
Will review later, what I use this for is:

 * n Workers need to perform an action at most once/twice/thrice per period
 * Out of 1000s of workers a subset n should be able to do something concurrently

For the first I set a stream up like this with a max messages of 1,2 or 3 per subject and this discard policy and a max age of the period.  So before doing work you put a message in, if it worked - you do the work.  If it failed you pass.  Real world use case I want to make cron jobs reliable over a pool of servers but only 1 of the pool must run a job, so bucket of 1 max, 59 minutes max age and so one of the cron hosts will run it.

For the 2nd I do the same but once I am done doing the work I remove the message I put in freeing up my slot for the next worker.  There is a max age on the stream so that workers who crash before removing their reservation will eventually timeout preventing starvation. Real world use is 10s of thousands of machiens need to download a file from Artifactory, I dont want to DDOS it, so I have a 100 slot bucket - so 100 at a time download from Artifactory.

**Comment by jnmoyne** [comment]
Good to read about your use cases, I may add them to the "that's not the only way you can use this" part in the conclusion.

**Comment by jnmoyne** [comment]
@gcolliso didn't know if there is some other file to modify to add the post to the blog besides creating a new .md file?

**Comment by gcolliso** [comment]
@jnmoyne Could you add an 'About the Author' section at the end?

**Review by derekcollison — APPROVED** [review]
LGTM

**Review by ripienaar — APPROVED** [review]
LGTM

**Comment by bruth** [comment]
@jnmoyne I stumbled upon this PR. Really well done and great use case to highlight! I am going to take another pass before Monday to fix a few grammatical things, but this is very cool and looking forward to writing up a complementary example for this.

**Comment by jnmoyne** [comment]
Thanks, will wait until you push grammar fixes and merge afterwards.

File: content/blog/new-per-subject-discard-policy.md
Comment (bruth) [inline content/blog/new-per-subject-discard-policy.md:95]:
```suggestion
And indeed, if you have experience with Kafka, this may look to you just like the "keep at least one message per key" log compaction feature, but there is a big difference: JetStream is exactly *n* maximum number of messages per subject while log compaction only ensures that you will have *at least one* message per key.
```

File: content/blog/new-per-subject-discard-policy.md
Comment (bruth) [inline content/blog/new-per-subject-discard-policy.md:97]:
```suggestion
Besides the potential inefficiency (i.e. having to scan the whole stream to be sure to get the latest message for a specific key in the last message cache) you will not be able to use a streaming WAL functionality to enforce a strict per key limit, and to signal the client application publishing a message immediately when the publication would break the limit. This is why there is no concept of a `new` discard policy in distributed log streaming systems, they are designed and optimized for always accepting new messages or block until they can.
```

File: content/blog/new-per-subject-discard-policy.md
Comment (bruth) [inline content/blog/new-per-subject-discard-policy.md:102]:
```suggestion
A prerequisite is access to a JetStream-enabled NATS server (e.g. running `nats-server -js` locally) and having the `nats` CLI context pointing to it.
```

File: content/blog/new-per-subject-discard-policy.md
Comment (bruth) [inline content/blog/new-per-subject-discard-policy.md:209]:
```suggestion
To review, the new discard policy *discard-new-per-subject* enables the ability to have an 'infinite deduplication window' to avoid duplicated publications that can otherwise happen to some failure scenarios.
```

File: content/blog/new-per-subject-discard-policy.md
Comment (bruth) [inline content/blog/new-per-subject-discard-policy.md:211]:
```suggestion
This new ability is specific to JetStream due to its fundamental functional differences compared to other distributed streaming systems: it is subject-based addressing capable and has the features of a proper data store (rather than the features of an append-only log with compaction). The functionality is built-in to NATS server itself rather than something you have to implement and deploy separately (e.g. Kafka Streams processes) and requires an external data store (e.g. Redis). 
```

File: content/blog/new-per-subject-discard-policy.md
Comment (bruth) [inline content/blog/new-per-subject-discard-policy.md:214]:
```suggestion

- N workers need to perform an action at most once/twice/thrice per period
```

File: content/blog/new-per-subject-discard-policy.md
Comment (bruth) [inline content/blog/new-per-subject-discard-policy.md:215]:
```suggestion
- Given 1000s of workers on an N subset should be able to do something concurrently
```

**Review by bruth — CHANGES_REQUESTED** [review]

**Review by gcolliso — APPROVED** [review]
lgtm

File: content/blog/new-per-subject-discard-policy.md
Comment (jnmoyne) [reply content/blog/new-per-subject-discard-policy.md]: That JS has the concept of 'old' and 'new' discard policies, while log systems don't have the concept of a discard policy because they only can support the equivalent of the 'old' discard policy (because they always either accept the message to the stream of block until it can be accepted)

File: content/blog/new-per-subject-discard-policy.md
Comment (jnmoyne) [reply content/blog/new-per-subject-discard-policy.md]: You are right there's a problem with the sentence, corrected in the last commit

**Pull request merged**
